### PR TITLE
BX-1080 - Add op Sends to testing

### PR DESCRIPTION
**Pull request by BrodyHughes**
L2 sends have broken several times and due to no e2e coverage, we haven't caught it until QA does a pass. This PR adds `Hardhat Optimism` as a chain and then sets up a new test suite that checks a simple OP send.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -52,6 +52,7 @@
     "bundle": "yarn build && yarn zip",
     "// Runs tests": "",
     "anvil": "ALCHEMY_API_KEY=$(grep ALCHEMY_API_KEY .env | cut -d '=' -f2) && anvil --fork-url https://eth-mainnet.alchemyapi.io/v2/$ALCHEMY_API_KEY",
+    "anvil:optimism": "ALCHEMY_API_KEY=$(grep ALCHEMY_API_KEY .env | cut -d '=' -f2) && anvil --fork-url https://opt-mainnet.g.alchemy.com/v2/$ALCHEMY_API_KEY",
```
Comment (brunobar79) [inline package.json:55]:
```suggestion
    "anvil:optimism": "OPTIMISM_MAINNET_RPC=$(grep OPTIMISM_MAINNET_RPC .env | cut -d '=' -f2) && anvil --fork-url $OPTIMISM_MAINNET_RPC",
```

File: src/entries/popup/pages/send/index.tsx [hunk 2]
```diff
@@ -189,6 +191,12 @@ export function Send() {
     saveSendTokenAddressAndChain,
   } = usePopupInstanceStore();
 
+  const returnedChainId = chainIdToUse(
```
Comment (brunobar79) [inline src/entries/popup/pages/send/index.tsx:194]: activeChainId maybe?

File: package.json [hunk 3]
```diff
@@ -60,6 +61,7 @@
     "vitest:serial": "./scripts/e2e-serial-tests.sh",
     "vitest:swap": "./scripts/e2e-serial-tests.sh 'swap'",
     "vitest:send": "./scripts/e2e-serial-tests.sh 'send'",
+    "vitest:optimism": "./scripts/e2e-op-serial-tests.sh 'optimismTransactions'",
```
Comment (brunobar79) [inline package.json:64]:
```suggestion
    "vitest:send:optimism": "./scripts/e2e-op-serial-tests.sh 'optimismTransactions'",
```

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -298,6 +298,39 @@ jobs:
...
+        continue-on-error: true
+        uses: nick-fields/retry@v2
+        with:
+          timeout_minutes: 25
+          max_attempts: 2
+          command: |
+            export BROWSER=chrome
+            export OS=linux
+            export CHROMIUM_BIN=$(find chrome -type f -name 'chrome')
+            yarn vitest:optimism --retry=5
```
Comment (brunobar79) [inline .github/workflows/build.yml:324]:
```suggestion
            yarn vitest:send:optimism --retry=5
```

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -298,6 +298,39 @@ jobs:
       - name: Fail if any tests failed
         if: steps.ChromeE2ESend.outcome == 'failure'
         run: exit 1
+  chrome-l2-e2e-send:
```
Comment (brunobar79) [inline .github/workflows/build.yml:301]:
```suggestion
  chrome-optimism-e2e-send:
```

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -298,6 +298,39 @@ jobs:
...
+    env:
+      DISPLAY: :0
+      VITEST_SEGFAULT_RETRY: 3
+    steps:
+      - uses: actions/checkout@v3
+      - uses: ./.github/actions/chromeTestsSetup
+        with:
+          gh-access-token: ${{ secrets.DOTENV_GITHUB_ACCESS_TOKEN }}
+      - name: Run L2 e2e send (Chrome)
+        id: ChromeL2E2ESend
```
Comment (brunobar79) [inline .github/workflows/build.yml:314]:
```suggestion
        id: ChromeOPE2ESend
```

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -298,6 +298,39 @@ jobs:
...
+        with:
+          timeout_minutes: 25
+          max_attempts: 2
+          command: |
+            export BROWSER=chrome
+            export OS=linux
+            export CHROMIUM_BIN=$(find chrome -type f -name 'chrome')
+            yarn vitest:optimism --retry=5
+      - name: Upload deps artifacts
+        if: steps.ChromeL2E2ESend.outcome == 'failure'
```
Comment (brunobar79) [inline .github/workflows/build.yml:326]:
```suggestion
        if: steps.ChromeOPE2ESend.outcome == 'failure'
```

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -298,6 +298,39 @@ jobs:
...
+        uses: actions/upload-artifact@v3
+        with:
+          name: screenshots
+          path: screenshots/
+      - name: Fail if any tests failed
+        if: steps.ChromeL2E2ESend.outcome == 'failure'
```
Comment (brunobar79) [inline .github/workflows/build.yml:332]:
```suggestion
        if: steps.ChromeOPE2ESend.outcome == 'failure'
```

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -298,6 +298,39 @@ jobs:
       - name: Fail if any tests failed
```
Comment (brunobar79) [inline .github/workflows/build.yml:298]: Ignore the spot where the comment is, but we need to add `chrome-optimism-e2e-send` to `cleanup > needs` array

**Review by brunobar79 — APPROVED** [review]
Looks great!

File: package.json (hunk 1 above)
Comment (BrodyHughes) [reply package.json:55]: ^ need waynes input on which is correct

File: src/entries/popup/hooks/approveAppRequest/useApproveAppRequestValidations.ts [hunk 10]
```diff
@@ -17,8 +17,14 @@ export const useApproveAppRequestValidations = ({
   chainId: ChainId;
   selectedGas?: GasFeeParams | GasFeeLegacyParams;
 }) => {
-  const { connectedToHardhat } = useConnectedToHardhatStore();
-  const chainIdToUse = connectedToHardhat ? ChainId.mainnet : chainId;
+  const { connectedToHardhat, connectedToHardhatOp } =
+    useConnectedToHardhatStore.getState();
```
Comment (estebanmino) [inline src/entries/popup/hooks/approveAppRequest/useApproveAppRequestValidations.ts:21]: `getState` is not needed here

File: src/entries/popup/pages/send/index.tsx [hunk 11]
```diff
@@ -82,7 +83,8 @@ export function Send() {
   const isMyWallet = (address: Address) =>
     allWallets?.some((w) => w.address === address);
 
-  const { connectedToHardhat } = useConnectedToHardhatStore();
+  const { connectedToHardhat, connectedToHardhatOp } =
+    useConnectedToHardhatStore.getState();
```
Comment (estebanmino) [inline src/entries/popup/pages/send/index.tsx:87]: `getState` not needed here

File: src/core/state/currentSettings/connectedToHardhat.ts [hunk 12]
```diff
@@ -13,10 +16,15 @@ export const connectedToHardhatStore = createStore<ConnectedToHardhatState>(
...
+
+    connectedToHardhatOp: false,
+    setConnectedToHardhatOp: (connectedToHardhatOp) => {
+      set({ connectedToHardhatOp });
+    },
   }),
   {
     persist: {
-      name: 'connectedToHardhat',
+      name: 'connectedToHardhatStore',
```
Comment (estebanmino) [inline src/core/state/currentSettings/connectedToHardhat.ts:27]: why the change of name? 

i believe don't use the `Store` word at the end for store names

File: src/core/resources/assets/userAssets.ts [hunk 13]
```diff
@@ -272,6 +275,38 @@ export async function parseUserAssets({
     }, {});
     parsedAssetsDict[ChainId.mainnet] = newMainnetAssets;
   }
+  if (connectedToHardhatOp) {
```
Comment (estebanmino) [inline src/core/resources/assets/userAssets.ts:278]: could we add this stuff to line 242 instead?
most of this code is copy pasted and the only difference i see it the provider we use to get balances (?)

```
if (connectedToHardhat || connectedToHardhatOp) {
   const hardhatChainId = connectedToHardhat ? ChainId.hardhat  : ChainId.hardhatOP
    const provider = getProvider({ chainId: hardhatChainId });
    // force checking for ETH if connected to hardhat
    const mainnetAssets = parsedAssetsDict[hardhatChainId];
    mainnetAssets[ETH_MAINNET_ASSET.uniqueId] = ETH_MAINNET_ASSET;
    if (process.env.IS_TESTING === 'true') {
      mainnetAssets[USDC_MAINNET_ASSET.uniqueId] = USDC_MAINNET_ASSET;
      mainnetAssets[DAI_MAINNET_ASSET.uniqueId] = DAI_MAINNET_ASSET;
    }
    const mainnetBalanceRequests = Object.values(mainnetAssets).map(
      async (asset) => {
        if (asset.chainId !== hardhatChainId) return asset;
...
```

something like that

File: src/core/state/currentSettings/connectedToHardhat.ts (hunk 12 above)
Comment (BrodyHughes) [reply src/core/state/currentSettings/connectedToHardhat.ts:27]: probs a naming mistake then. will fix.

File: e2e/helpers.ts [hunk 14]
```diff
@@ -687,13 +687,10 @@ export async function getOnchainBalance(addy: string, contract: string) {
...
   const provider = getDefaultProvider('http://127.0.0.1:8545');
   const blockData = await provider.getBlock('latest');
-  const txn = await provider.getTransaction(blockData.transactions[0]);
-  const txnData = txn.wait();
-
-  // transactionResponse.wait.status returns '1' if txn is successful
-  // it returns '0' if the txn is a failure
-  const txnStatus = (await txnData).status === 1 ? 'success' : 'failure';
-
+  const txnReceipt = await provider.getTransactionReceipt(
```
Comment (greg-schrammel) [inline e2e/helpers.ts:690]: idk if it matters in this testing env, but `txnReceipt` will be `null` if the tx is `pending`

File: src/entries/popup/hooks/approveAppRequest/useApproveAppRequestValidations.ts (hunk 10 above)
Comment (greg-schrammel) [reply src/entries/popup/hooks/approveAppRequest/useApproveAppRequestValidations.ts:21]: using `getState` would get the state at the time the hook was called, but won't react to changes
using the hook `useConnectedToHardhatStore` will make all dependent components rerender when this state changes

File: src/core/resources/assets/userAssets.ts [hunk 15]
```diff
@@ -219,7 +220,9 @@ export async function parseUserAssets({
   chainIds: ChainId[];
   currency: SupportedCurrencyKey;
 }) {
-  const { connectedToHardhat } = connectedToHardhatStore.getState();
+  const { connectedToHardhat, connectedToHardhatOp } =
+    useConnectedToHardhatStore.getState();
```
Comment (greg-schrammel) [inline src/core/resources/assets/userAssets.ts:224]: but for ex here you use `.getState()` instead of the hook because it's a `parseUserAssets` function, it doesn't need to react to these changes and would probably break the rule of hooks because its called from react query and not directly from a component/other hook

Comment (greg-schrammel) [reply src/core/resources/assets/userAssets.ts:224]: not my best explanation happy to elaborate on slack if you want lol

File: e2e/helpers.ts (hunk 14 above)
Comment (BrodyHughes) [reply e2e/helpers.ts:690]: honestly fine as long as it returns 'failure' for this function

File: src/core/resources/assets/userAssets.ts [hunk 16]
```diff
@@ -236,41 +236,57 @@ export async function parseUserAssets({
...
+    const newParsedAssetsByUniqueId = await Promise.all(balanceRequests);
+    const newAssets = newParsedAssetsByUniqueId.reduce<
       Record<string, ParsedUserAsset>
     >((acc, parsedAsset) => {
       acc[parsedAsset.uniqueId] = parsedAsset;
       return acc;
     }, {});
-    parsedAssetsDict[ChainId.mainnet] = newMainnetAssets;
+
+    parsedAssetsDict[ChainId.mainnet] = newAssets;
```
Comment (estebanmino) [inline src/core/resources/assets/userAssets.ts:289]: i think you should use `mainnetOrOptimismChainId` ?

Comment (estebanmino) [reply src/core/resources/assets/userAssets.ts:289]: instead of mainnet

File: src/core/state/currentSettings/connectedToHardhat.ts [hunk 17]
```diff
@@ -22,4 +30,4 @@ export const connectedToHardhatStore = createStore<ConnectedToHardhatState>(
   },
 );
 
-export const useConnectedToHardhatStore = create(connectedToHardhatStore);
```
Comment (estebanmino) [inline src/core/state/currentSettings/connectedToHardhat.ts:25]: oh the naming thing was just for the `persist` param, we do use the `Store` end on all of these

File: src/entries/popup/components/SideChainExplainer.tsx [hunk 18]
```diff
@@ -8,7 +8,7 @@ import {
   ExplainerSheetProps,
 } from './ExplainerSheet/ExplainerSheet';
 
-type SideChain = Exclude<ChainId, 1 | 5 | 1337>;
+type SideChain = Exclude<ChainId, 1 | 5 | 1337 | 1338>;
```
Comment (estebanmino) [inline src/entries/popup/components/SideChainExplainer.tsx:11]: not part of your pr but could we do something like

```
type SideChain = Exclude<ChainId, ChainId.mainnet | ChainId.goerli | ChainId.hardhat  | ChainId.hardhatOp >;
```
?
